test/test_run.py

```diff
@@ -346,17 +350,50 @@ def test_safe_shell_exec_interrupts_on_event(self):
 
         sleep = 10
         start = time.time()
-        self._do_test_safe_shell_exec('sleep {}'.format(sleep), 143, '', 'Terminated\n', interrupt)
+        self._do_test_safe_shell_exec('sleep {}'.format(sleep), 143, '', '', interrupt)
         duration = time.time() - start
 
         self.assertGreaterEqual(duration, 1.0)
         self.assertLess(duration, 2.0 + safe_shell_exec.GRACEFUL_TERMINATION_TIME_S, 'sleep should not finish')
         self.assertGreater(sleep, 2.0 + safe_shell_exec.GRACEFUL_TERMINATION_TIME_S, 'sleep should allow for GRACEFUL_TERMINATION_TIME_S')
 
+    def test_safe_shell_exec_interrupts_on_parent_shutdown(self):
+        sleep = 20
+        parent_script = os.path.join(os.path.dirname(__file__), 'data/run_safe_shell_exec.py')
+        child_script = os.path.join(os.path.dirname(__file__), 'data/sleep.py')
+
+        with temppath() as logfile:
+            # It's important that this executes in an entirely different interpreter with as little shared
+            # state as possible, to avoid issues with the semaphore tracker.
+            cmd = ' '.join([sys.executable, parent_script, child_script, str(sleep), logfile])
+            p = subprocess.Popen(cmd, shell=True)
+
+            # Wait until the child script has written its PID to the logfile
+            wait(lambda: os.path.exists(logfile), timeout=5)
+
+            parent = psutil.Process(p.pid)
+            with open(logfile, 'r') as f:
+                child_pid = int(f.read())
+            child = psutil.Process(child_pid)
+
+            self.assertTrue(parent.is_running())
+            self.assertTrue(child.is_running())
+
+            # Hard kill the parent process
+            parent.kill()
+            parent.wait(timeout=safe_shell_exec.GRACEFUL_TERMINATION_TIME_S)
+            p.wait()
+
+            # Child process will exit when pipe breaks
+            child.wait(timeout=safe_shell_exec.GRACEFUL_TERMINATION_TIME_S)
```

Comment:
in worst case `child` can take up to `2 * safe_shell_exec.GRACEFUL_TERMINATION_TIME_S`:

- `psutil.wait_procs(p.children(), timeout=GRACEFUL_TERMINATION_TIME_S)`
- `p.wait(timeout=GRACEFUL_TERMINATION_TIME_S)`

Plus give it an extra second so the test does not out-race the middleman.